### PR TITLE
BGL:  doxygen doc for concepts

### LINKED ISSUE #1246: One cannot search anything about the BGL in the search box of the doc
The documentation of specializations of `boost::graph_traits`, for 4.8.1, are in [the documentation page of the group `PkgBGLTraits`](http://doc.cgal.org/4.8.1/BGL/group__PkgBGLTraits.html).

<s>
## 1st, a small documentation bug

<s>There is a bug, in a description of two iterator types, for `Polyhedron_3`:

| Member | Value |
| --- | --- |
| halfedge_iterator | Polyhedron_3::Halfedge_iterator |
| face_iterator | Polyhedron_3::Face_iterator |

It cannot be true, because the value types of the iterators of the polyhedron are not descriptors but directly the pointed type.
</s> (**update:** see the dedicated bug entry #1251).
</s>
## 2nd, one cannot search anything about the BGL in the search box of the doc

In the Doxygen search box, one cannot search "graph_traits", "faces", "num_faces", and so on... That is a real usability issue. I think all the files `BGL/include/CGAL/boost/graph/graph_traits_*.h` should be processed by Doxygen, with a rough documentation that links to [the page of the group `PkgBGLTraits`](http://doc.cgal.org/4.8.1/BGL/group__PkgBGLTraits.html).

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

Regroup the tables for valid expressions of the various graph concepts on one page, as it is done  on [boost.org](http://www.boost.org/doc/libs/1_64_0/libs/graph/doc/graph_concepts.html), and document the free function with `\relates` to the pages of the individual concepts. 

This is WIP in order to get feedback. 

## Release Management

* Affected package(s): BGL
* Issue(s) solved (if any): fix #1246

### CODE REVIEW

**Comment by lrineau** [comment]
That PR is against `master`. As it is, it will not be integrated into CGAL-4.10.

File: BGL/doc/BGL/Concepts/HalfedgeGraph.h [hunk 1]
```diff
@@ -23,43 +23,68 @@ A model of `HalfedgeGraph` must have the interior property `vertex_point` attach
...
-`opposite(h, g)`                        | `halfedge_descriptor`                                                        | The halfedge with source and target swapped.
-`source(h,g)`                           | `vertex_descriptor`                                                          | The source vertex of `h`.
-`target(h,g)`                           | `vertex_descriptor`                                                          | The target vertex of `h`.
-`next(h, g)`                            | `halfedge_descriptor`                                                        | The next halfedge around its face.
-`prev(h, g)`                            | `halfedge_descriptor`                                                        | The previous halfedge around its face.
-`boost::graph_traits<G>::%null_halfedge()` | `halfedge_descriptor`                                                     | Returns a special halfedge that is not equal to any other halfedge.
-
-\cgalHeading{Invariants}
-
-`halfedge(edge(h,g))==h`
```
Comment (sloriot) [inline BGL/doc/BGL/Concepts/HalfedgeGraph.h]: You removed this invariant that is very important.

File: BGL/doc/BGL/Concepts/FaceListGraph.h [hunk 2]
```diff
@@ -9,33 +9,25 @@ the requirement for traversal of all faces in a graph.
...
 \cgalHasModel `CGAL::Surface_mesh`
 
-\cgalHeading{Notations}
-
-<dl>
-<dt>`G`</dt> 	<dd>A type that is a model of `FaceListGraph`.</dd>
-<dt>`g`</dt> 	<dd>An object of type `G`.</dd>
-</dl>
+*/
+class FaceListGraph{};
```
Comment (sloriot) [inline BGL/doc/BGL/Concepts/FaceListGraph.h]: For this class, all related functions are using `const FaceListGraph&` while for `FaceGraph` it is only `FaceGraph&`.
Also, the difference with the previous documentation was that the constness was not specified. One solution is to use `FaceGraph&` and say somewhere that `FaceGraph` might be `const FaceGraph_bis `.

File: BGL/doc/BGL/Concepts/EdgeListGraph.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+\cgalHasModel `CGAL::Surface_mesh`
+
+*/
+class EdgeListGraph{};
+
+
+/*! \relates EdgeListGraph
+ * returns an iterator range over all edges.
+ */
+std::pair<boost::graph_traits<EdgeListGraph>::edge_iterator,
```
Comment (sloriot) [inline BGL/doc/BGL/Concepts/EdgeListGraph.h]: All the free functions of all the concepts are missing `template <class FooBarGraph>`

File: BGL/doc/BGL/PackageDescription.txt [hunk 4]
```diff
@@ -1,33 +1,193 @@
...
-///  FaceListGraph -> FaceGraph -> HalfedgeGraph [ arrowhead="open", label="refines" ];
-///  MutableFaceGraph -> MutableHalfedgeGraph [ arrowhead="open", label="refines" ];
-///  MutableFaceGraph -> FaceGraph [ arrowhead="open", label="refines" ];
-///  HalfedgeListGraph -> HalfedgeGraph [ arrowhead="open", label="refines" ];
-///}
-///\enddot
+/*! \defgroup PkgBGLConcepts Concepts
+ \ingroup PkgBGL
+
+ We extend the Boost Graph Library (\sc{Bgl} for short) with a set of new concepts.
```
Comment (sloriot) [inline BGL/doc/BGL/PackageDescription.txt]: Actually `EdgeListGraph` is already a concept in the BGL. I guess you want to mention it to get links to free functions, but in that case you better mention `VertexAndEdgeListGraph` to also get vertex related functions.

File: BGL/doc/BGL/Concepts/HalfedgeGraph.h [hunk 5]
```diff
@@ -23,43 +23,78 @@ A model of `HalfedgeGraph` must have the interior property `vertex_point` attach
...
-
-
 */
 class HalfedgeGraph {};
+
+/*! \relates HalfedgeGraph
+returns the edge corresponding to halfedges `h` and `opposite(h)`.
```
Comment (sloriot) [inline BGL/doc/BGL/Concepts/HalfedgeGraph.h]: `opposite(h)` -> `opposite(h,g)`

File: BGL/doc/BGL/PackageDescription.txt [hunk 6]
```diff
@@ -99,6 +99,8 @@ Valid Expression                        | Returns
 `prev(h, g)`                            | `halfedge_descriptor`                       | The previous halfedge around its face.
 `boost::graph_traits<G>::%null_halfedge()` | `halfedge_descriptor`                    | Returns a special halfedge that is not equal to any other halfedge.
 
+The `HalfedgeGraph`has the invariant  `halfedge(edge(h,g))==h`.
```
Comment (lrineau) [inline BGL/doc/BGL/PackageDescription.txt]: typo: missing space between `HalfedgeGraph` and "has".

File: BGL/doc/BGL/Concepts/HalfedgeGraph.h (hunk 1 above)
Comment (lrineau) [inline BGL/doc/BGL/Concepts/HalfedgeGraph.h]: @sloriot, https://github.com/CGAL/cgal/pull/2086#discussion_r116725808: 

    \cgalHeading{Invariants}

    `halfedge(edge(h,g))==h`
> You removed this invariant that is very important.

@afabri: Have you re-added the invariant?

File: BGL/doc/BGL/PackageDescription.txt (hunk 6 above)
Comment (lrineau) [inline BGL/doc/BGL/PackageDescription.txt]: ...  Well, here I can see that the invariant was indeed re-added...

File: BGL/doc/BGL/Concepts/HalfedgeGraph.h (hunk 1 above)
Comment (lrineau) [inline BGL/doc/BGL/Concepts/HalfedgeGraph.h]: I answer myself: yes, the invariant was re-added. See https://github.com/CGAL/cgal/pull/2086#discussion_r118283443

**Pull request merged**